pitop/processing/algorithms/ball_detect.py

```diff
@@ -0,0 +1,231 @@
+                        # store the values to be used as the most likely ball in frame
+                        ball_center = (int(x), int(y))
+                        ball_radius = int(radius)
+
+    return ball_center, ball_radius
+
+
+def process_frame_for_ball(frame, colours=("red",), image_return_format: str = "PIL", scale_factor=0.5):
+    cv_frame = ImageFunctions.convert(frame, format="OpenCV")
+    resized_frame = scale_frame(cv_frame, scale=scale_factor)
+
+    if len(colours) > 3:
+        raise ValueError('Cannot pass more than three colours.')
+
+    ball_centers = {}
+    ball_radii = {}
+    ball_angles = {}
+    ball_finds = {}
+    for colour in colours:
+        mask = colour_filter(resized_frame, colour=colour, return_binary_mask=True, image_format="OpenCV",
+                             image_return_format="OpenCV")
+
+        contours = cv2.findContours(mask.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+        contours = imutils.grab_contours(contours)  # fixes problems with OpenCV changing their protocol
+
+        ball_center, ball_radius = find_most_likely_ball(contours, colour, resized_frame)
+
+        detection_points[colour].appendleft(ball_center)
+
+        if ball_center is not None:
+            # draw the circle and centroid on the frame, then update the list of tracked points
+            cv2.circle(resized_frame, ball_center, ball_radius, (0, 255, 255), 2)
+            cv2.circle(resized_frame, ball_center, 5, draw_colour[colour], -1)
+
+            # Reposition centre so (0, 0) is in the middle for the user.
+            # Keep scale as 1 as user sees scaled down 320x240 image in Further
+            ball_center = centroid_reposition(ball_center, 1, resized_frame)
+            ball_angle = get_control_angle(ball_center, resized_frame)
+
+            ball_centers[colour] = ball_center
+            ball_radii[colour] = ball_radius
+            ball_angles[colour] = ball_angle
+            ball_finds[colour] = True
+        else:
+            ball_centers[colour] = None
+            ball_radii[colour] = None
+            ball_angles[colour] = None
+            ball_finds[colour] = False
+
+        for i in range(1, len(detection_points[colour])):
+            # if either of the tracked points are None, ignore them
+            if detection_points[colour][i - 1] is None or detection_points[colour][i] is None:
+                continue
+            # otherwise, compute the thickness of the line and draw the connecting lines
+            thickness = int(np.sqrt(BUFFER_LENGTH / float(i + 1)))
+
+            cv2.line(resized_frame, detection_points[colour][i - 1], detection_points[colour][i],
+                     draw_colour[colour], thickness)
+
+    if image_return_format.lower() != 'opencv':
+        robot_view = ImageFunctions.convert(resized_frame, format="PIL")
+    else:
+        robot_view = resized_frame
+
+    if len(colours) == 1:
```

Comment:
@jcapona when you get a minute, let's discuss this ball_detect return structure. It's a bit odd but it gave me what I wanted for the Further challenge - it almost definitely needs a refactor. Potentially a Ball class with `@dataclass` decorate that includes red, green and blue balls. The only issue is if they only pass one colour to find, it should be accessible just with e.g `ball.radius` instead of `ball.red.radius`